File: memind-clients/python/src/memind/types/memory.py
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from __future__ import annotations

from typing import Any

from pydantic import Field

from memind._models import MemindModel
from memind.types.common import Strategy
from memind.types.message import Message, RawContentValue


class ExtractMemoryRequest(MemindModel):
    user_id: str
    agent_id: str
    raw_content: RawContentValue
    source_client: str | None = None


class AddMessageRequest(MemindModel):
    user_id: str
    agent_id: str
    message: Message
    source_client: str | None = None


class CommitMemoryRequest(MemindModel):
    user_id: str
    agent_id: str
    source_client: str | None = None


class ExtractMemoryResponse(MemindModel):
    status: str
    raw_data_ids: list[str] = Field(default_factory=list)
    item_ids: list[int] = Field(default_factory=list)
    insight_ids: list[int] = Field(default_factory=list)
    insight_pending: bool = False
    duration_millis: int | None = None
    error_message: str | None = None


class AddMessageResponse(MemindModel):
    triggered: bool
    result: ExtractMemoryResponse | None = None


class RetrieveMemoryRequest(MemindModel):
    user_id: str
    agent_id: str
    query: str
    strategy: Strategy
    trace: bool | None = None
    scope: str | None = None
    categories: list[str] | None = None
    time_range: TimeRange | None = None
    metadata_filter: MetadataFilter | None = None
    include: RetrieveIncludeOptions | None = None


class TimeRange(MemindModel):
    field: str | None = None
    from_: str | None = Field(default=None, alias="from")
    to: str | None = None


class MetadataCondition(MemindModel):
    path: str
    op: str
    value: Any | None = None


class MetadataFilter(MemindModel):
    all: list[MetadataCondition] | None = None
    any: list[MetadataCondition] | None = None
    not_: list[MetadataCondition] | None = Field(default=None, alias="not")


class RetrieveIncludeOptions(MemindModel):
    raw_data_metadata: bool | None = None
    raw_data_segment: bool | None = None


class RawDataQueryIncludeOptions(MemindModel):
    segment: bool | None = None
    metadata: bool | None = None


class QueryMemoryItemsRequest(MemindModel):
    user_id: str
    agent_id: str
    scope: str | None = None
    categories: list[str] | None = None
    source_clients: list[str] | None = None
    raw_data_types: list[str] | None = None
    time_range: TimeRange | None = None
    metadata_filter: MetadataFilter | None = None
    limit: int | None = None
    cursor: str | None = None


class QueryMemoryRawDataRequest(MemindModel):
    user_id: str
    agent_id: str
    types: list[str] | None = None
    source_clients: list[str] | None = None
    time_range: TimeRange | None = None
    metadata_filter: MetadataFilter | None = None
    include: RawDataQueryIncludeOptions | None = None
    limit: int | None = None
    cursor: str | None = None


class RetrievedItem(MemindModel):
    id: str
    text: str
    vector_score: float = 0.0
    final_score: float = 0.0
    occurred_at: str | None = None
    category: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class RetrievedInsight(MemindModel):
    id: str
    text: str
    tier: str | None = None


class RetrievedRawData(MemindModel):
    raw_data_id: str
    caption: str | None = None
    max_score: float = 0.0
    item_ids: list[str] | None = None
    type: str | None = None
    source_client: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    start_time: str | None = None
    end_time: str | None = None
    created_at: str | None = None


class StageView(MemindModel):
    stage: str | None = None
    tier: str | None = None
    method: str | None = None
    status: str | None = None
    input_count: int | None = None
    candidate_count: int | None = None
    result_count: int | None = None
    degraded: bool = False
    skipped: bool = False
    started_at: str | None = None
    duration_millis: int | None = None
    attributes: dict[str, Any] | None = None
    candidates: list[dict[str, Any]] | None = None


class MergeView(MemindModel):
    input_count: int = 0
    output_count: int = 0
    deduplicated_count: int = 0
    source_count: int = 0
    status: str | None = None


class FinalView(MemindModel):
    strategy: str | None = None
    status: str | None = None
    item_count: int = 0
    insight_count: int = 0
    raw_data_count: int = 0
    evidence_count: int = 0


class RetrievalTraceView(MemindModel):
    trace_id: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    truncated: bool | None = None
    stages: list[StageView] = Field(default_factory=list)
    merge: MergeView | None = None
    final_results: FinalView | None = None


class RetrieveMemoryResponse(MemindModel):
    status: str | None = None
    items: list[RetrievedItem] = Field(default_factory=list)
    insights: list[RetrievedInsight] = Field(default_factory=list)
    raw_data: list[RetrievedRawData] = Field(default_factory=list)
    evidences: list[str] = Field(default_factory=list)
    strategy: str | None = None
    query: str | None = None
    trace: RetrievalTraceView | None = None


class MemoryItem(MemindModel):
    id: str
    text: str
    scope: str | None = None
    category: str | None = None
    type: str | None = None
    raw_data_id: str | None = None
    raw_data_type: str | None = None
    source_client: str | None = None
    occurred_at: str | None = None
    observed_at: str | None = None
    created_at: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class QueryMemoryItemsResponse(MemindModel):
    items: list[MemoryItem] = Field(default_factory=list)
    next_cursor: str | None = None


class MemoryRawData(MemindModel):
    id: str
    type: str | None = None
    source_client: str | None = None
    caption: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    segment: dict[str, Any] | None = None
    start_time: str | None = None
    end_time: str | None = None
    created_at: str | None = None


class QueryMemoryRawDataResponse(MemindModel):
    raw_data: list[MemoryRawData] = Field(default_factory=list)
    next_cursor: str | None = None

File: memind-clients/typescript/src/types/memory.ts
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import type { Strategy } from './common.js'
import type { MessageValue, RawContentValue } from './message.js'

export type MetadataCondition = {
  path: string
  op: 'eq' | 'in' | 'exists' | 'missing' | 'contains' | string
  value?: unknown
}

export type MetadataFilter = {
  all?: MetadataCondition[]
  any?: MetadataCondition[]
  not?: MetadataCondition[]
}

export type TimeRange = {
  field?: string
  from?: string
  to?: string
}

export type RetrieveIncludeOptions = {
  rawDataMetadata?: boolean
  rawDataSegment?: boolean
}

export type RawDataQueryIncludeOptions = {
  segment?: boolean
  metadata?: boolean
}

export type ExtractMemoryRequest = {
  userId: string
  agentId: string
  rawContent: RawContentValue
  sourceClient?: string
}

export type AddMessageRequest = {
  userId: string
  agentId: string
  message: MessageValue
  sourceClient?: string
}

export type CommitMemoryRequest = {
  userId: string
  agentId: string
  sourceClient?: string
}

export type RetrieveMemoryRequest = {
  userId: string
  agentId: string
  query: string
  strategy: Strategy
  trace?: boolean
  scope?: string
  categories?: string[]
  timeRange?: TimeRange
  metadataFilter?: MetadataFilter
  include?: RetrieveIncludeOptions
}

export type QueryMemoryItemsRequest = {
  userId: string
  agentId: string
  scope?: string
  categories?: string[]
  sourceClients?: string[]
  rawDataTypes?: string[]
  timeRange?: TimeRange
  metadataFilter?: MetadataFilter
  limit?: number
  cursor?: string
}

export type QueryMemoryRawDataRequest = {
  userId: string
  agentId: string
  types?: string[]
  sourceClients?: string[]
  timeRange?: TimeRange
  metadataFilter?: MetadataFilter
  include?: RawDataQueryIncludeOptions
  limit?: number
  cursor?: string
}

export type ExtractMemoryResponse = {
  status: string
  rawDataIds: string[]
  itemIds: number[]
  insightIds: number[]
  insightPending: boolean
  durationMillis?: number
  errorMessage?: string
}

export type AddMessageResponse = {
  triggered: boolean
  result?: ExtractMemoryResponse
}

export type RetrievedItem = {
  id: string
  text: string
  vectorScore: number
  finalScore: number
  occurredAt?: string
  category?: string
  metadata?: Record<string, unknown>
}

export type RetrievedInsight = {
  id: string
  text: string
  tier?: string
}

export type RetrievedRawData = {
  rawDataId: string
  caption?: string
  maxScore: number
  itemIds?: string[]
  type?: string
  sourceClient?: string
  metadata?: Record<string, unknown>
  startTime?: string
  endTime?: string
  createdAt?: string
}

export type StageView = {
  stage?: string
  tier?: string
  method?: string
  status?: string
  inputCount?: number
  candidateCount?: number
  resultCount?: number
  degraded: boolean
  skipped: boolean
  startedAt?: string
  durationMillis?: number
  attributes?: Record<string, unknown>
  candidates?: Record<string, unknown>[]
}

export type MergeView = {
  inputCount: number
  outputCount: number
  deduplicatedCount: number
  sourceCount: number
  status?: string
}

export type FinalView = {
  strategy?: string
  status?: string
  itemCount: number
  insightCount: number
  rawDataCount: number
  evidenceCount: number
}

export type RetrievalTraceView = {
  traceId?: string
  startedAt?: string
  completedAt?: string
  truncated?: boolean
  stages: StageView[]
  merge?: MergeView
  finalResults?: FinalView
}

export type RetrieveMemoryResponse = {
  status?: string
  items: RetrievedItem[]
  insights: RetrievedInsight[]
  rawData: RetrievedRawData[]
  evidences: string[]
  strategy?: string
  query?: string
  trace?: RetrievalTraceView
}

export type MemoryItem = {
  id: string
  text: string
  scope?: string
  category?: string
  type?: string
  rawDataId?: string
  rawDataType?: string
  sourceClient?: string
  occurredAt?: string
  observedAt?: string
  createdAt?: string
  metadata?: Record<string, unknown>
}

export type QueryMemoryItemsResponse = {
  items: MemoryItem[]
  nextCursor?: string
}

export type MemoryRawData = {
  id: string
  type?: string
  sourceClient?: string
  caption?: string
  metadata?: Record<string, unknown>
  segment?: Record<string, unknown>
  startTime?: string
  endTime?: string
  createdAt?: string
}

export type QueryMemoryRawDataResponse = {
  rawData: MemoryRawData[]
  nextCursor?: string
}
